system-addon/lib/MessageCenterRouter.jsm

```diff
@@ -0,0 +1,204 @@
+ *
+ * Note: This is written as a constructor rather than just a plain object
+ * so that it can be more easily unit tested.
+ */
+class _MessageCenterRouter {
+  constructor(initialState = {}) {
+    this.initialized = false;
+    this.messageChannel = null;
+    this._state = Object.assign({
+      currentId: null,
+      blockList: {},
+      messages: {}
+    }, initialState);
+    this.onMessage = this.onMessage.bind(this);
+  }
+
+  get state() {
+    return this._state;
+  }
+
+  set state(value) {
+    throw new Error("Do not modify this.state directy. Instead, call this.setState(newState)");
+  }
+
+  setState(callbackOrObj) {
+    const newState = (typeof callbackOrObj === "function") ? callbackOrObj(this.state) : callbackOrObj;
+    this._state = Object.assign({}, this.state, newState);
+    return new Promise(resolve => {
+      this._onStateChanged(this.state);
+      resolve();
+    });
+  }
+
+  _onStateChanged(state) {
+    this.messageChannel.sendAsyncMessage(OUTGOING_MESSAGE_NAME, {type: "ADMIN_SET_STATE", data: state});
+  }
+
+  async sendNextMessage(target, id) {
+    let message;
+    await this.setState(state => {
+      message = getRandomItemFromArray(state.messages.filter(item => item.id !== state.currentId && !state.blockList[item.id]));
+      return {currentId: message ? message.id : null};
+    });
+    if (message) {
+      target.sendAsyncMessage(OUTGOING_MESSAGE_NAME, {type: "SET_MESSAGE", data: message});
+    } else {
+      target.sendAsyncMessage(OUTGOING_MESSAGE_NAME, {type: "CLEAR_MESSAGE"});
+    }
+  }
+
+  async clearMessage(target, id) {
+    if (this.state.currentId === id) {
+      await this.setState({currentId: null});
+    }
+    target.sendAsyncMessage(OUTGOING_MESSAGE_NAME, {type: "CLEAR_MESSAGE"});
+  }
+
+  async onMessage({data: action, target}) {
+    switch (action.type) {
+      case "CONNECT_UI_REQUEST":
+      case "GET_NEXT_MESSAGE":
+        await this.sendNextMessage(target);
+        break;
+      case "BLOCK_MESSAGE_BY_ID":
+        await this.setState(state => {
+          const newState = Object.assign({}, state.blockList);
+          newState[action.data.id] = true;
+          return {blockList: newState};
+        });
+        this.clearMessage(target, action.data.id);
+        break;
+      case "UNBLOCK_MESSAGE_BY_ID":
+        await this.setState(state => {
+          const newState = Object.assign({}, state.blockList);
+          delete newState[action.data.id];
+          return {blockList: newState};
+        });
+        break;
+      case "ADMIN_CONNECT_STATE":
+        target.sendAsyncMessage(OUTGOING_MESSAGE_NAME, {type: "ADMIN_SET_STATE", data: this.state});
+        break;
+    }
+  }
+
+  /**
+   * init - Initializes the MessageRouter.
+   * It is ready when it has been connected to a RemotePageManager instance.
+   *
+   * @param {RemotePageManager} channel a RemotePageManager instance
+   * @memberof _MessageCenterRouter
+   */
+  init(channel) {
```

Comment:
Can this be higher up in the class? It makes it difficult to look at the constructor and `init` at the same time.